File: Using_CSharp/behavioral/Ducks/Ducks/Duck.cs
using System;

namespace Ducks
{
	public abstract class Duck {

    // Maintain references to the each behavior interface:
    public IFlyBehavior flyBehavior;
    public IQuackBehavior quackBehavior;

    
		public Duck() { }

    //-------------------------------------------------------------------------
    public abstract void Display();

    //-------------------------------------------------------------------------
    public void PerformFly() {
      flyBehavior.Fly();
    }

    //-------------------------------------------------------------------------
    public void PerformQuack() {
      quackBehavior.QuackBehavior();
    }

    //-------------------------------------------------------------------------
    public void Swim() {
      Console.WriteLine("All ducks float, even decoys");
    }
	}
}

File: using_python/behavioral/ducks/duck.py
#!/usr/bin/env

# The duck base class.

class Duck:

    # Instance variables that hold a reference to a specific behavior at runtime.
    fly_behavior = None
    quack_behavior = None

    def set_fly_behavior(self, fly_behavior):
        self.fly_behavior = fly_behavior

    def set_quack_behavior(self, quack_behavior):
        self.quack_behavior = quack_behavior

    def display(self):
        raise NotImplementedError

    def perform_fly(self):
        self.fly_behavior.fly()

    def perform_quack(self):
        self.quack_behavior.quack()

    def swim(self):
        print("All ducks float, even decoys")
